Add an abstraction over device' fwnode

vvarma opened the pull request:
I was conflicted about naming the module `fwnode` when the functions it implements are from `linux/property.h`, yet i feel `fwnode` provides more clarity about what the module provides. I have (currently) implemented only the `endpoints` method which provides a rust interface to traversing the tree. There were no changes required in `binding_helper.rs` since `linux/property.h` is already included in some of the other headers included.

bjorn3 commented on rust/kernel/fwnode.rs:
```diff
@@ -0,0 +1,111 @@
+pub struct NodeIterator {
+    handle: DeviceNode,
+    curr_node: Option<Arc<Node>>,
+    iter_fn: NodeIteratorFn,
```
I think `NodeIterator` should use `next_endpoint` directly without a function pointer in between.

bjorn3 commented on rust/kernel/fwnode.rs:
```diff
@@ -0,0 +1,111 @@
+///
+/// The pointer is valid.
+pub struct Node {
+    pub(crate) ptr: *mut bindings::fwnode_handle,
```
I think this should store `UnsafeCell<bindings::fwnode_handle>` and be `#[repr(transparent)]` directly and implement `AlwaysRefCounted`. Then you can return an `ARef<Node>` from `NodeIterator`. See for example https://github.com/Rust-for-Linux/linux/blob/729c03011c60382030520f84904a85e8a5b3cfb9/rust/kernel/file.rs#L113-L179 You can then remove the Send, Sync and Drop impls.

vvarma replied:
thanks @bjorn3 for the review! Is it correct to reason that this is not required for DeviceNode since it's ptr is not public? And hence its safe to rely on rust's reference counting completely.

bjorn3 replied:
`fwnode_handle` is reference counted too, so `DeviceNode` should probably do the same. Also `dev_fwnode` seems to return a pointer without incrementing reference count. As such I think `DeviceNode::from` should return `&'a DeviceNode` where `'a` is the lifetime of the device.

bjorn3 replied:
Also I don't know if you endpoints of nodes can have endpoints themself, but if so maybe `DeviceNode` should be merged into `Node`.

vvarma replied:
> if you endpoints of nodes can have endpoints themself,

I took your recommendation and merged `DeviceNode` and `Node`. Based on the [documentation](https://www.kernel.org/doc/html/latest/firmware-guide/acpi/dsd/graph.html) endpoints should not have endpoints themselves. However, calling `endpoints()` returns an Iterator which may be empty as well, so should be safe for consumption. 

> dev_fwnode seems to return a pointer without incrementing reference count.

I have added a `fwnode_handle_get` call to ensure that the refcount is incremented before passing it to the ARef. Since I am explicitly incrementing the refcount in `Node::from` I am not associating the lifetime of `Node` with the device. Am i missing anything?

bjorn3 replied:
> > dev_fwnode seems to return a pointer without incrementing reference count.
> 
> I have added a `fwnode_handle_get` call to ensure that the refcount is incremented before passing it to the ARef. Since I am explicitly incrementing the refcount in `Node::from` I am not associating the lifetime of `Node` with the device. Am i missing anything?

Should be fine now.

bjorn3 commented on rust/kernel/fwnode.rs:
```diff
@@ -0,0 +1,88 @@
+    }
+}
+
+/// Implements the Iterator trait to iterate the device's endpoints given the `DeviceNode`
```
Here and elsewhere there are still references to DeviceNode.

vvarma replied:
My bad! I have fixed them now.

bjorn3 reviewed (approved):
Based on what I know about this interface, it looks good. I'm not sure if `Node` should be called this or something closer to `fwnode_handle` though like `FwNodeHandle`.

vvarma commented:
> Based on what I know about this interface, it looks good. I'm not sure if `Node` should be called this or something closer to `fwnode_handle` though like `FwNodeHandle`.

Since the module name is called `fwnode` wont it be repetitive to call it `fwnode::FwNodeHandle` ? However, if thats the convention I could rename it to that.

I have also added two methods `parent` and `children`. I have reverted to passing a function to the `NodeIterator` (like i had in an earlier version of this PR), to reuse the iterator for the `endpoints` and the `children` methods, both of which have the same pattern.
